### ISSUE TITLE
Ethercalc doesn't transform/rebase concurrent inserts and deletes correctly (or at all?)

### ISSUE THREAD

**Issue by marcelklehr** [audreyt/ethercalc#327]
1. Open a new sheet and enter some data.
2. Open the sheet in a second window
3. Try to simultaneously change the value of a cell in one window, while inserting/deleting a column in the other window.

Expected result: The intended cell is changed and the documents are in a cosistent state.

Actual result: In the window where the column was inserted the wrong cell is changed.

The reason for this is that afaict the edits are not rebased (or transformed in OT language).

**Comment by audreyt** [comment]
That is correct. Our OT fork is here: https://github.com/iclems/firecalc — we have yet to merge it back.

**Comment by marcelklehr** [comment]
Apparently they haven't implemented this either: https://github.com/iclems/firecalc/blob/master/lib/operation.js#L40

I've started work on [ot-socialcalc](https://github.com/marcelklehr/ot-socialcalc).

**Comment by audreyt** [comment]
Wonderful — Let there be 💡!

**Comment by eddyparkinson** [comment]
Marcel,  Thanks for looking at this. - I always value a 2nd pair of eyes. Would you like me to review a design doc? 
Eddy.
 

    On Monday, 11 April 2016, 21:17, Marcel Klehr <notifications@github.com> wrote:
 

 Apparently they haven't implemented this either: https://github.com/iclems/firecalc/blob/master/lib/operation.js#L40I've started work on ot-socialcalc.—

**Comment by marcelklehr** [comment]
@eddyparkinson Sure, I'll try and roughly sketch the design here.

The ot type is defined on the format of snapshots that socialcalc itself uses. An operation for the consumer of the ot type actually consists of many small operation objects that each translate to a socialcalc command. Operations are deserialized from socialcalc commands and can be serialized into commands again. The purpose of having operation objects is simpler/more concise transformation logic. Operations are applied on a snapshot internally by loading the snapshot into a socialcalc instance, serializing the operations to a command string, feeding it to the socialcalc instance and serializing the new spreadsheet state into a snapshot again.

**Comment by eddyparkinson** [comment]
@marcelklehr
1) ot type issue:  yes, the ot type is the core of the problem. The ot type includes row and column.This causes a problem when rows and cols are inserted.

 e.g commands: 
   "set B8 text t william"   "insertrow A1"
These commands cause a problem. If  "set B8 text t william" is executed on machine A and at the same time "insertrow A1" is executed on machine B.   And then this will be followed by  "insertrow A1" on machine A and  "set B8 text t william"  on machine B. 
To see this in action, in the browser, add a break point in the file ethercalc.js  in the method SocialCalc.ExecuteSheetCommand

2) Hope this helps, it took me a long time to learn my way around the major functions. 

   

 On Tuesday, 12 April 2016, 9:11, 唐鳳 <notifications@github.com> wrote:
 

 Wonderful — Let there be 💡 !—

**Comment by marcelklehr** [comment]
Yes, that's the exact problem. The solution is to transform the commands against each other, so that they fit and the resulting document states are consistent.

See for example https://github.com/marcelklehr/ot-socialcalc/blob/master/index.js#L128 for how the commands should be transformed in your example.

**Comment by eddyparkinson** [comment]
Issue:    Line 378 & Line 431  should they both follow the same pattern of doing DeleteCol ?  as one is doing InsertCol  
If I have some time later I will have a better look, but looks like you might be done before I get the time.
Eddy. ---
Eddy Parkinson PhD
Software Engineering and Logistics
Mobile 0406 962 958 

New ideas pass through three periods: 1) It can`t be done. 2) It 
probably can be done, but it`s not worth doing. 3) I knew it was a good 
idea all along! - Arthur C. Clarke
 

    On Wednesday, 13 April 2016, 21:40, Marcel Klehr <notifications@github.com> wrote:
 

 Yes, that's the exact problem. The solution is to transform the commands against each other, so that they fit and the resulting document states are consistent.See for example https://github.com/marcelklehr/ot-socialcalc/blob/master/index.js#L128 for how the commands should be transformed in your example.—

**Comment by marcelklehr** [comment]
Thank you for taking the time to look into this!

Lines 378 and 431 shouldn't be the same because they cover different transformations. (Variable `op` is the *other* operation, the one that is being transformed against, while `this` is the operation being transformed.) In the first case InsertCol is transformed against DeleteCol, hence the outcome should always be an InsertCol. In the second case DeleteCol a is transformed against DeleteCol b, hence the outcome should be DeleteCol a'. So, even though they're both transformed against DeleteCol, the outcome is different because the operations being transformed are of different types.

**Comment by marcelklehr** [comment]
So, I've run into an issue: Cut&Paste. Ethercalc currently broadcasts copy paste commands like everything else, but that results in there being only one buffer for all users, which is not what we want, IMO. We'd need to translate `copy`/`cut`&`paste`s into normal `set`s, which cannot be done inside ot-socialcalc, because these commands of course don't contain the actual data.

**Comment by audreyt** [comment]
Yes, we should do that. One way is to transform, at `emit` (`player.ls`) level, a cut/copy into `loadclipboard` equivalents.

**Comment by eddyparkinson** [comment]
1) Normal "set" sounds like a good solution. 
2) Another option: -    Clone the loadclipboard  command. Clone both the "loadclipboard" and "paste" command to create a new "pastedata" command. Add a command like "pastedata" that combines "loadclipboard"  and "paste" into a single command. And create the "pastedata" command so it does not change the clipboard.  But would this clash "insertcol B1"  and "pastedata A1 .... " ?
example loadclipboard "loadclipboard version\c1.5\ncell\cA1\ct\cd\ncell\cB1\ct\ce\nsheet\cc\c2\cr\c1\ncopiedfrom\cA1\cB1\n"


I think either could be made to work. Not sure which is better. 
---
Eddy Parkinson PhD
Software Engineering and Logistics

New ideas pass through three periods: 1) It can`t be done. 2) It 
probably can be done, but it`s not worth doing. 3) I knew it was a good 
idea all along! - Arthur C. Clarke
 

    On Sunday, 17 April 2016, 10:40, Marcel Klehr <notifications@github.com> wrote:
 

 So, I've run into an issue: Cut&Paste. Ethercalc currently broadcasts copy paste commands like everything else, but that results in there being only one buffer for all users, which is not what we want, IMO. We'd need to translate copy/cut&pastes into normal sets, which cannot be done inside ot-socialcalc, because these commands of course don't contain the actual data.—

**Comment by marcelklehr** [comment]
Ok, so I've tried to wrap my head around this. I think having operations that operate on blocks is probably a bad idea, because we need to guarantee consistency when transforming against column and row insertions/deletetions. It's probably simpler (and saner) to operate on cell level only.

Some questions off the top of my head: does `set` allow ranges? Does `set column/row` actually set attributes on the row/column, or does the editor translate that into single cell mutations (the latter would be another breakage of consistency)?

**Comment by marcelklehr** [comment]
As a background info on why I'm doing this: I'm working on [hive.js](http://hivejs.org), a collaboration platform for which I've written a package integrating SocialCalc. If you're curios, be sure to drop by in the [gitter chat room for hive](https://gitter.im/hivejs/chat), if you like :)

**Comment by eddyparkinson** [comment]
>Does set column/row actually set attributes on the row/column
yes

// hide
set G hide yes 
set 6:6 hide yes
set 9 hide yes

// unhide
set G hide
set 6 hide

// width
set H width 100

//Height:
There is a set height - not sure if the UI has access, but the command exists.


Only hide and width/height - as far as I can tell from the code.

>does the editor translate that into single cell mutations 
No.  I have looked at this in the past and I concluded it impacts the entire row/col.
But it has been a while since I looked at this.

**Comment by eddyparkinson** [comment]
Re:  hive.js, & OT
 Hive.js is impressive. 
 I had a look at hive.js when I first saw your post, 

  Is there an OT for data syncing via an API, For example, syncing customer order details in one database  with accounting information in another database?

**Comment by eddyparkinson** [comment]
Look related - Merged cells - insert row/col  https://github.com/audreyt/ethercalc/issues/341

**Comment by marcelklehr** [comment]
@eddyparkinson re data syncing API: I'm not sure if this is what you're looking for, but [gulf](https://github.com/marcelklehr/gulf) is an abstraction over OT that you could use for something like this.

re merged cells: Ah, I'll comment on the issue... thanks
